sql: add active user txn information to crdb_internal

rohany opened the pull request:
Fixes #46055.

Release justification: low risk change to existing functionality

Release note (sql change): This PR:
* Adds new internal tables `crdb_internal.node_transactions`
and `crdb_internal.cluster_transactions.`
* These tables contain some metadata about active user transactions.
* Adds the column `txn_id` to the `crdb_internal.node_queries` and
  `crdb_internal.cluster_queries` tables. These fields represent
  the transaction ID of each query in each row.

rohany committed:
sql: add active user txn information to crdb_internal

Fixes #46055.

Release justification: low risk change to existing functionality

Release note (sql change): This PR:
* Adds new internal tables `crdb_internal.node_transactions`
and `crdb_internal.cluster_transactions.`
* These tables contain some metadata about active user transactions.
* Adds the column `txn_id` to the `crdb_internal.node_queries` and
  `crdb_internal.cluster_queries` tables. These fields represent
  the transaction ID of each query in each row.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46206)
<!-- Reviewable:end -->

rohany commented:
Clarifiying some things before writing tests:
* Is this the data you're looking for @ajwerner?
* Is there some interaction with this and rollbacks/savepoints? I.E. will the transaction id recorded here need to change if a statement is retried or rolled back to a savepoint?
* What is the best way to test this? Is there some way to verify that the recorded txn UUID points to a valid txn?
* Does this need some sort of version gating? or are blank fields for `txn_start` and `txn_id` reasonable return values in mixed version clusters?

awoods187 commented:
nice quick fix!

On Tue, Mar 17, 2020 at 1:30 PM Rohan Yadav <notifications@github.com>
wrote:

> Clarifiying some things before writing tests:
>
>    - Is this the data you're looking for @ajwerner
>    <https://github.com/ajwerner>?
>    - Is there some interaction with this and rollbacks/savepoints? I.E.
>    will the transaction id recorded here need to change if a statement is
>    retried or rolled back to a savepoint?
>    - What is the best way to test this? Is there some way to verify that
>    the recorded txn UUID points to a valid txn?
>    - Does this need some sort of version gating? or are blank fields for
>    txn_start and txn_id reasonable return values in mixed version
>    clusters?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/46206#issuecomment-600200633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJ7F7ZOZ2SXFXAWKIYVRBLRH6XUZANCNFSM4LNVGG3A>
> .
>


-- 
Andy Woods
Senior Product Manager, Cockroach Labs
andy@cockroachlabs.com

ajwerner commented:
> Is this the data you're looking for @ajwerner?

Indeed.

> Is there some interaction with this and rollbacks/savepoints? I.E. will the transaction id recorded here need to change if a statement is retried or rolled back to a savepoint?

I don't believe so, no. Transaction retries use the same ID. They increment the epoch. There are some cases (one I know if posted below) where we will create a new transaction upon retry but I'm not sure it's worth dealing with them here. 

https://github.com/cockroachdb/cockroach/blob/733563a521a9323f77411a609af1ea019073c6c9/pkg/sql/conn_executor_exec.go#L578-L603

> What is the best way to test this? Is there some way to verify that the recorded txn UUID points to a valid txn?

You could use tracing. The batch requests will have it. 

> Does this need some sort of version gating? or are blank fields for txn_start and txn_id reasonable return values in mixed version clusters?

This is a good question. I'm not sure I'm qualified to answer it.

andreimatei commented:
Instead of this, how about a different virtual table for ongoing SQL txns? It'd be good to know open transactions that are not currently running a query too.

rohany commented:
Alright, I added a test and added some null handling in a mixed cluster setting.

I'm open to another virtual table, but why would we need to choose between this and that?

andreimatei commented:
Cause if we had a different table, I think we'd just link to that, and not put the txn start in the output of SHOW QUERIES.

rohany commented:
Duh, valid. Let me give adding that a shot.

rohany commented:
Any requests for what should be in the transactions table? So far i have txn id, start time, node id, and session id.

rohany commented:
Updated this to create a new virtual table, and populate that virtual table with open transaction information.

I'm a bit worried this might be too large of a change to merge in the stability period though.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @knz, and @rohany)

___
*[docs/generated/sql/bnf/stmt_block.bnf, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iP7VA3qGjHcNif248:-M2iP7VB9zqnJV3JaNNF:b1q2fgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/docs/generated/sql/bnf/stmt_block.bnf#L559)):*
> ```
> 	| 'SHOW' opt_compact 'KV' 'TRACE' 'FOR' 'SESSION'
> 
> show_transactions_stmt ::=
> ```

I don't think I'd introduce this `show transactions` just yet (or ever). Let's let the virtual tables mature for a while. Changing the schema of a `SHOW` seems like a more difficult decision than changing a virtual table.

Also it's not clear to me what the default should be - local transactions or cluster-wide.
___
*[pkg/server/serverpb/status.proto, line 429 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLOT14rwwSEs5QKxn:-M2iLOT14rwwSEs5QKxo:bom7e4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L429)):*
> ```protobuf
> 
> // ActiveUserTxn represents an in flight user transaction on some Session.
> message ActiveUserTxn {
> ```

I think "active" and "user" are unclear, and in any case don't belong in the type name (this message could describe an "inactive" or "non-user" transaction as well). I'd name it `TxnInfo`.
___
*[pkg/server/serverpb/status.proto, line 430 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLr7x9WJoLXccb5Yg:-M2iLr7x9WJoLXccb5Yh:b-tbs61o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L430)):*
> ```protobuf
> // ActiveUserTxn represents an in flight user transaction on some Session.
> message ActiveUserTxn {
>   string id = 1 [ (gogoproto.customname) = "ID" ];
> ```

Let's include the string representation of the underlying `kv.Txn`; it's gonna be really useful; I've hacked such a registry several time when debugging stuff. There's no `String()` method currently, but there should be - it should go through `txn.mu.sender` and eventually call `TxnCoordSender.mu.txn.String()`. 
___
*[pkg/server/serverpb/status.proto, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLL7m6tvji6kolCDP:-M2iLL7m6tvji6kolCDQ:b-e7xlvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L484)):*
> ```protobuf
>   repeated ActiveQuery active_queries = 5 [ (gogoproto.nullable) = false ];
>   // User Txns in progress on this session.
>   repeated ActiveUserTxn active_txns = 12 [ (gogoproto.nullable) = false ];
> ```

can there be more than one at a time?
___
*[pkg/sql/conn_executor.go, line 1079 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iNOtfDyH3EVL9RH75:-M2iNOtfDyH3EVL9RH76:bhdn3c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/conn_executor.go#L1079)):*
> ```Go
> 
> 		// ActiveUserTxns contains all user txns in flight.
> 		ActiveUserTxns map[uuid.UUID]*txnMeta
> ```

can there be more than one? I don't think you need any new state here at all. You have the current txn, if any, inside `txnState`, and you've got its start time in that `phaseTimes` thing. No?
___
*[pkg/sql/crdb_internal.go, line 938 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iOIvTBBXjzArlXSdV:-M2iOIvTBBXjzArlXSdW:bnfznif) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/crdb_internal.go#L938)):*
> ```Go
> const txnsSchemaPattern = `
> CREATE TABLE crdb_internal.%s (
> 	id STRING,               -- the unique ID of the transaction
> ```

nit: misaligned
___
*[pkg/sql/crdb_internal.go, line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iOfWa31shaBCi35Mw:-M2iOfWa31shaBCi35Mx:buj91rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/crdb_internal.go#L985)):*
> ```Go
> 			); err != nil {
> 				return err
> 			}
> ```

Are any errors here expected, and below? If not, let's panic instead and not return an error from this function.
As written, it seems like a funky interface - we're swallowing the errors that actually matter (the rpc errors), but we're still burdening the caller with handling returned errors for no good reason.


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @knz)

___
*[docs/generated/sql/bnf/stmt_block.bnf, line 559 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iP7VA3qGjHcNif248:-M2igjpI0t2m4fAnrLVv:bw547wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/docs/generated/sql/bnf/stmt_block.bnf#L559)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think I'd introduce this `show transactions` just yet (or ever). Let's let the virtual tables mature for a while. Changing the schema of a `SHOW` seems like a more difficult decision than changing a virtual table.

Also it's not clear to me what the default should be - local transactions or cluster-wide.
</blockquote></details>

that makes sense. I'll remove this (so much boilerplate added F)
___
*[pkg/server/serverpb/status.proto, line 429 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLOT14rwwSEs5QKxn:-M2ieFjH4aGTufaPb9L2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L429)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think "active" and "user" are unclear, and in any case don't belong in the type name (this message could describe an "inactive" or "non-user" transaction as well). I'd name it `TxnInfo`.
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 430 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLr7x9WJoLXccb5Yg:-M2ieElQ7WrOvIVQ0HOQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L430)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Let's include the string representation of the underlying `kv.Txn`; it's gonna be really useful; I've hacked such a registry several time when debugging stuff. There's no `String()` method currently, but there should be - it should go through `txn.mu.sender` and eventually call `TxnCoordSender.mu.txn.String()`. 
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLL7m6tvji6kolCDP:-M2iV_otAb44bKR69xYc:b-fus3v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L484)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can there be more than one at a time?
</blockquote></details>

answered below
___
*[pkg/sql/conn_executor.go, line 1079 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iNOtfDyH3EVL9RH75:-M2iVJQfDAR9L-e-ILtS:b-obzpog) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/conn_executor.go#L1079)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can there be more than one? I don't think you need any new state here at all. You have the current txn, if any, inside `txnState`, and you've got its start time in that `phaseTimes` thing. No?
</blockquote></details>

I'm not sure, you tell me! We record multiple active queries for a single `conn_executor` so shouldn't we be able to have multiple transactions open at the same time? If not, this makes life simpler.
___
*[pkg/sql/crdb_internal.go, line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iOfWa31shaBCi35Mw:-M2iUdyzA3gj_67gKSMJ:b2okz5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/crdb_internal.go#L985)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Are any errors here expected, and below? If not, let's panic instead and not return an error from this function.
As written, it seems like a funky interface - we're swallowing the errors that actually matter (the rpc errors), but we're still burdening the caller with handling returned errors for no good reason.
</blockquote></details>

I was cargo culting from the corresponding tables for `node_queries` and `cluster_queries`. I think errors are expected here (if we are collecting a debug zip when a node in the cluster is down). If i include a string version of the txn, then I can put the error there (this is similar to what the node queries table does -- it replaces the query text with the rpc error)


<!-- Sent from Reviewable.io -->

knz reviewed (changes requested):
I'm generally OK with the approach but the new functionality must be admin-only.

The reason for this is that the debugging details extracted with String can contain KV key data, which can contain PII details. It's unreasonable to show this information to any user without filter.

The alternative is to not populate the details column if the requesting user is non-admin.

> Reviewed 4 of 22 files at r1, 25 of 25 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-:-M2msgMO5hgSWJqzf2Hs:bvc2aph)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @rohany)

___
*[pkg/server/serverpb/status.proto, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLL7m6tvji6kolCDP:-M2mrNliDTnuTrLV70Rf:b-jt3vwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L484)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

answered below
</blockquote></details>

I would add the explanation as a comment on the field.
___
*[pkg/server/serverpb/status.proto, line 435 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2mr61V6JSfZxgCe6Gh:-M2mr61V6JSfZxgCe6Gi:b-toovm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/server/serverpb/status.proto#L435)):*
> ```protobuf
>     [ (gogoproto.nullable) = false, (gogoproto.stdtime) = true ];
> 
>   // txn_string is the string representation of the underlying kv.Txn.
> ```

1) maybe `txn_description` instead of `txn_string`

2) "[field name] is a text description of the underlying kv.Txn, intended for troubleshooting purposes"

(The current phrasing suggests it's possible to re-construct the txn object from the string. This would be incorrect.)
___
*[pkg/sql/conn_executor.go, line 1079 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iNOtfDyH3EVL9RH75:-M2mrSkg6nHmMFST4Lcu:b-qk0wtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/conn_executor.go#L1079)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I'm not sure, you tell me! We record multiple active queries for a single `conn_executor` so shouldn't we be able to have multiple transactions open at the same time? If not, this makes life simpler.
</blockquote></details>

@andreimatei we used to have concurrent query execution. That's why this complexity exists. I recall we removed that some time ago (v19.2 maybe?). If so, maybe we can collapse all this complexity throughout conn executor.
___
*[pkg/sql/crdb_internal.go, line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iOfWa31shaBCi35Mw:-M2mrmat7sfFyHggYHos:b-2xgy8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/crdb_internal.go#L985)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I was cargo culting from the corresponding tables for `node_queries` and `cluster_queries`. I think errors are expected here (if we are collecting a debug zip when a node in the cluster is down). If i include a string version of the txn, then I can put the error there (this is similar to what the node queries table does -- it replaces the query text with the rpc error)
</blockquote></details>

nit: @rohany please ignore Andrei when he says "let's panic". This was best practice back in v2.x. In the new world since v19.2, if there's an `error` return, you can use `return errors.AssertionFailedf(...)` with details.
___
*[pkg/sql/crdb_internal.go, line 950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2msc0H7B4KE29BYR1r:-M2msc0H7B4KE29BYR1s:b-9p1vzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/sql/crdb_internal.go#L950)):*
> ```Go
> 	schema:  fmt.Sprintf(txnsSchemaPattern, "node_transactions"),
> 	populate: func(ctx context.Context, p *planner, _ *DatabaseDescriptor, addRow func(...tree.Datum) error) error {
> 		req := p.makeSessionsRequest(ctx)
> ```

make this and the other new tables admin-only, thanks (see my comment at top)
___
*[pkg/sql/exec_util.go, line 1000 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2ms1_64WyEaGpjkGip:-M2ms1_64WyEaGpjkGiq:b-gzaktj) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/sql/exec_util.go#L1000)):*
> ```Go
> 	txnStart time.Time
> 
> 	// The string representation of this transaction.
> ```

see my comment above. "string" is too generic a word. Try "description" or something better.


<!-- Sent from Reviewable.io -->

knz commented:
>  Transaction retries use the same ID. 

That's not generally true. "Simple retries" when the txn itself detects a conflict keep the ID,

However if the txn gets aborted *by another txn* (e.g. of a higher priority):
- the kv txn will get trashed.
- however, the *SQL txn* is preserved (the client can handle using `ROLLBACK TO SAVEPOINT cockroach_restart` and retry)

So a single SQL txn can have multiple successive kv txns.

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @knz, and @rohany)

___
*[pkg/server/serverpb/status.proto, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iLL7m6tvji6kolCDP:-M2nOrSiDA9sHWMxnUdb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/server/serverpb/status.proto#L484)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would add the explanation as a comment on the field.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1079 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iNOtfDyH3EVL9RH75:-M2nNxds2vvO42dagXtC:b-zaim0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/conn_executor.go#L1079)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

@andreimatei we used to have concurrent query execution. That's why this complexity exists. I recall we removed that some time ago (v19.2 maybe?). If so, maybe we can collapse all this complexity throughout conn executor.
</blockquote></details>

removed the extra state for txns
___
*[pkg/sql/crdb_internal.go, line 950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2msc0H7B4KE29BYR1r:-M2nGmQc6mr7u46PTl5Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/sql/crdb_internal.go#L950)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make this and the other new tables admin-only, thanks (see my comment at top)
</blockquote></details>

Done.
___
*[pkg/sql/exec_util.go, line 1000 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2ms1_64WyEaGpjkGip:-M2nGYlw5pfas3K66fIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/sql/exec_util.go#L1000)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

see my comment above. "string" is too generic a word. Try "description" or something better.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @knz, and @rohany)

___
*[pkg/server/serverpb/status.proto, line 435 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2mr61V6JSfZxgCe6Gh:-M2nMzc_BmqmhSb2i5Fn:b-6qjp8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/58c728660a281f470a617168ca40b97d26460166/pkg/server/serverpb/status.proto#L435)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) maybe `txn_description` instead of `txn_string`

2) "[field name] is a text description of the underlying kv.Txn, intended for troubleshooting purposes"

(The current phrasing suggests it's possible to re-construct the txn object from the string. This would be incorrect.)
</blockquote></details>

my preference would be simply `info`
___
*[pkg/sql/conn_executor.go, line 1079 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iNOtfDyH3EVL9RH75:-M2nN725D4qZP6CXQNuZ:bvkfovg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/conn_executor.go#L1079)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

@andreimatei we used to have concurrent query execution. That's why this complexity exists. I recall we removed that some time ago (v19.2 maybe?). If so, maybe we can collapse all this complexity throughout conn executor.
</blockquote></details>

yeah, the multiple queries in parallel stuff used to be for `returning nothing` queries. We don't execute them in parallel any more. Please get rid of the ability to track multiple ones if you feel like it.
Even then, there could have been a single txn at a time.
___
*[pkg/sql/crdb_internal.go, line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2iOfWa31shaBCi35Mw:-M2nOvKl4YXkFW3y1gCK:b-51bnnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4306e5334ec2cda229ba701b2989e6488b96648f/pkg/sql/crdb_internal.go#L985)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: @rohany please ignore Andrei when he says "let's panic". This was best practice back in v2.x. In the new world since v19.2, if there's an `error` return, you can use `return errors.AssertionFailedf(...)` with details.
</blockquote></details>

I'd rather not just be ignored :)
I was talking strictly about errors from `addRow()`. Reading the code more, I think I was wrong though; looks like that function can actually fail for dynamic reasons - memory limits. So there's nothing to assert.

The point is moot, but I was arguing for panicing on assertion failures here. My policy is that deciding when to panic and when to return an assertion error is a calculus on how dynamic the conditions for failing the assertion are and how much effort is it for the code to accommodate that error. If the function with the assertion wouldn't otherwise need to return an error (which I thought was the case here), in most cases I would add an error return just for assertions.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
LGTM but see the comment about the new field in `show queries`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-:-M2nUDE-CuUS5MHOWd7P:b8r6365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @knz, and @rohany)

___
*[pkg/sql/conn_executor_exec.go, line 1193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2nTg9r2HcL52aPxdQf:-M2nTg9sDF79sa2Mw61I:b-v5kc9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c4bb333de515a958fd3c0ed306019cc073b9c2a/pkg/sql/conn_executor_exec.go#L1193)):*
> ```Go
> 	_, hidden := stmt.AST.(tree.HiddenFromShowQueries)
> 	qm := &queryMeta{
> 		txnID:         ex.state.mu.txn.ID(),
> ```

Since we're not adding a SHOW porcelain for the transactions, I'd rather not add the transaction ID in `show queries`. Seems weird to do this kind of crossing between the `show` world and the internal tables.
What do you think?


<!-- Sent from Reviewable.io -->

rohany commented:
> Since we're not adding a SHOW porcelain for the transactions, I'd rather not add the transaction ID in show queries. Seems weird to do this kind of crossing between the show world and the internal tables. What do you think?

Well this whole thing started because @ajwerner wanted to see the txn id / start time in show queries, so we should have a way to link query -> txn info. It seems fine to cross this boundary (at least in this case) because we just aren't sure yet what all should go into the new show statement.

knz commented:
> Well this whole thing started because @ajwerner wanted to see the txn id / start time in show queries,

A *programmatic* requirement needs to be answered exclusively in a vtable. SHOW is meant for human users and not meant for use by automation.

So it's ok to provide what andrew needs in the vtable without revealing it in the SHOW output.

rohany commented:
Yeah, that makes sense. I can keep txn id in the internal *queries table but remove it from the show command.

knz reviewed (approved):
LGTM modulo proper identification of the insufficient privilege error

> Reviewed 6 of 6 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-:-M2niEO4-D3EUs5EFnH3:b-83audt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @rohany)

___
*[pkg/sql/crdb_internal.go, line 984 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-M2ni8X1-gnMshnpGvwq:-M2ni8X1-gnMshnpGvwr:b-tlhr6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3c8a14923e66409746a65e9e8b524855d2b0990/pkg/sql/crdb_internal.go#L984)):*
> ```Go
> func (p *planner) makeSessionsRequest(ctx context.Context) serverpb.ListSessionsRequest {
> 	req := serverpb.ListSessionsRequest{Username: p.SessionData().User}
> 	if err := p.RequireAdminRole(ctx, "list sessions"); err == nil {
> ```

can you check the `err` object more specifically.


<!-- Sent from Reviewable.io -->

rohany commented:
> can you check the err object more specifically.

sorry, I'm not sure what you have in mind

knz commented:
`errors.Is` with the proper reference variable, or `pgerror.GetPGCode`

rohany commented:
But that variable is being compared to nil, not an error string (sorry i think i' missing something basic here)

knz reviewed (approved):
please ignore my previous comment about err - I was mistaken

> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46206#-:-M2oHybf-001klcynvPS:b-d86mwe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @rohany)



<!-- Sent from Reviewable.io -->

rohany commented:
bors r+

tftr!
